pkg/controller/nodes/task/catalog/datacatalog/datacatalog.go

```diff
@@ -269,9 +256,83 @@ func (m *CatalogClient) Put(ctx context.Context, key catalog.Key, reader io.Outp
+
+	artifactDataList := make([]*datacatalog.ArtifactData, 0, len(outputs.Literals))
+	for name, value := range outputs.Literals {
+		artifactData := &datacatalog.ArtifactData{
+			Name:  name,
+			Value: value,
+		}
+		artifactDataList = append(artifactDataList, artifactData)
+	}
+
+	tagName, err := GenerateArtifactTagName(ctx, inputs)
+	if err != nil {
+		logger.Errorf(ctx, "Failed to generate artifact tag name for key %+v, dataset %+v and execution %+v, err: %+v", key, datasetID, metadata, err)
+		return catalog.Status{}, err
+	}
+
+	updateArtifactRequest := &datacatalog.UpdateArtifactRequest{
+		Dataset:     datasetID,
+		QueryHandle: &datacatalog.UpdateArtifactRequest_TagName{TagName: tagName},
+		Data:        artifactDataList,
+	}
+	resp, err := m.client.UpdateArtifact(ctx, updateArtifactRequest)
+	if err != nil {
+		logger.Errorf(ctx, "Failed to update artifact for key %+v, dataset %+v and execution %+v, err: %v", key, datasetID, metadata, err)
+		return catalog.Status{}, err
+	}
+
+	tag := &datacatalog.Tag{
+		Name:       tagName,
+		Dataset:    datasetID,
+		ArtifactId: resp.GetArtifactId(),
+	}
+
+	source, err := GetSourceFromMetadata(GetDatasetMetadataForSource(metadata.TaskExecutionIdentifier), GetArtifactMetadataForSource(metadata.TaskExecutionIdentifier), key.Identifier)
+	if err != nil {
+		return catalog.Status{}, fmt.Errorf("failed to get source from metadata. Error: %w", err)
+	}
+
+	logger.Debugf(ctx, "Successfully updated artifact with ID %v and %d outputs for key %+v, dataset %+v and execution %+v", tag.ArtifactId, len(artifactDataList), key, datasetID, metadata)
+	return catalog.NewStatus(core.CatalogCacheStatus_CACHE_POPULATED, EventCatalogMetadata(datasetID, tag, source)), nil
+}
+
+// Put stores the result of a task execution as a cached Artifact and associates it with the data by tagging it with
+// the hash of the input values.
+// The CatalogClient will ensure a dataset exists for the Artifact to be created. A Dataset represents the
+// project/domain/name/version of the task executed.
+// If overwrite is set to true, Put will replace the ArtifactData of an existing Artifact with the new task execution
+// output.
+// Otherwise, CatalogClient will create an Artifact tagged with the input value hash and store the provided execution data.
+func (m *CatalogClient) Put(ctx context.Context, key catalog.Key, reader io.OutputReader, metadata catalog.Metadata, overwrite bool) (catalog.Status, error) {
+	// Ensure dataset exists, idempotent operations. Populate Metadata for later recovery
+	datasetID, err := m.CreateDataset(ctx, key, GetDatasetMetadataForSource(metadata.TaskExecutionIdentifier))
+	if err != nil {
+		return catalog.Status{}, err
+	}
+
+	inputs, outputs, err := m.prepareInputsAndOutputs(ctx, key, reader)
+	if err != nil {
+		return catalog.Status{}, err
+	}
+
+	// Overwrite existing artifact
+	if overwrite {
+		return m.UpdateArtifact(ctx, key, datasetID, inputs, outputs, metadata)
+	}
```

Comment:
IIUC if the user specifies the `skipCache` flag then we always call `Put` with the `overwrite = true` which then calls this `UpdateArtifact` function. In the datacatalog implementation it looks like this fails if the artifact does not exist. Is this the intended functionality? It seems to me that if the cache does not exist for the task then it should work the same (ie. successful) whether the `skipCache` flag is set or not. Thoughts?

Reply:
Good point, yes. I think it'll probably still be most consistent if the datacatalog implementation fails if the artifact does not exist, but we should just handle this case here (check the returned error, ignore anything indicating the artifact was not found).
Alternatively, we could do a lookup to see if data has been found before trying to remove that, but that would introduce the overhead of one additional call, so I'd rather just handle the error silently if that sounds acceptable to you.

Reply:
I think that sounds great! The only other idea I thought was to add an `upsert` flag to the update call which allows this to be an atomic operation. I think it makes more sense to keep datacatalog ops very simple (ie. Put, Get, Update). We just need to make sure that if there is a race condition on the `Insert` after the `Update` fails then we ignore a `AlreadyExists` error.